### PR TITLE
Implement skin support for (most) hitsounds and (osu!) hitcircles

### PR DESCRIPTION
**Pull request by peppy**
- Adds basic hitsound replacement (doesn't check ruleset prefixes, yet)
- Adds basic hitcircle skinnability (osu! ruleset)

also

- [x] Depends on #2103
- [x] Depends on ppy/osu-framework#1410

As `skin.ini` is not yet parsed, depending on the skin it may not yet look as expected. skin.ini parsing will be implemented in the next PR.

I have visual tests in the works for testing skinning, will push to this branch when ready. For now testing can be done by loading an `.osk` manually.

### CODE REVIEW

File: osu.Game/Skinning/SkinManager.cs [hunk 1]
```diff
@@ -30,13 +34,37 @@ public List<SkinInfo> GetAllUsableSkins()
...
 
         private SkinStore store;
 
-        public SkinManager(Storage storage, DatabaseContextFactory contextFactory, IIpcHost importHost)
+        public SkinManager(Storage storage, DatabaseContextFactory contextFactory, IIpcHost importHost, AudioManager audio)
             : base(storage, contextFactory, new SkinStore(contextFactory, storage), importHost)
         {
+            this.audio = audio;
+
+            CurrentSkinInfo.ValueChanged += info => { CurrentSkin.Value = GetSkin(info); };
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinManager.cs]: unnecessary braces

File: osu.Game/Skinning/SkinManager.cs [hunk 2]
```diff
@@ -30,13 +34,37 @@ public List<SkinInfo> GetAllUsableSkins()
...
+            CurrentSkin.ValueChanged += skin =>
+            {
+                if (skin.SkinInfo != CurrentSkinInfo.Value)
+                    throw new InvalidOperationException($"Setting {nameof(CurrentSkin)}'s value directly is not supported. Use {nameof(CurrentSkinInfo)} isntead.");
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinManager.cs]: instead* typo

File: osu.Game/Skinning/LegacySkin.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        private readonly TextureStore textures;
+
+        private readonly SampleManager samples;
+
+        public LegacySkin(SkinInfo skin, IResourceStore<byte[]> storage, AudioManager audioManager)
+            : base(skin)
+        {
+            var audioStore = new ResourceStore<byte[]>(storage);
+
+            samples = audioManager.GetSampleManager(audioStore);
```
Comment (smoogipoo) [inline osu.Game/Skinning/LegacySkin.cs]: Isn't this the same as `GetSampleManager(storage)`? Is there a reason we have a separate resourcestore here?

File: osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs [hunk 4]
```diff
@@ -82,32 +83,41 @@ protected DrawableHitObject(HitObject hitObject)
...
 
-                    Samples.Add(channel);
+                        if (channel == null) return;
+
+                        Samples.Add(channel);
+                    }
                 }
+
+                skin.ValueChanged += loadSamples;
+                skin.BindTo(skins.CurrentSkin);
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs]: No way, I thought you were going to abstract this out and have SkinManager return you the bindables... We shouldn't be storing local skin bindables like this.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs]: temporary implementation

File: osu.Game/Skinning/LegacySkin.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Skinning/LegacySkin.cs]: no, got left over from some previous code

File: osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs]: wait, not temporary. this is intentional. i don't want the SkinManager to be handling anything but skins.

Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs]: As discussed, let's add a SkinnableSample to handle this logic for us, in a similar way to SkinnableDrawable.

File: osu.Game/Skinning/SkinReloadableDrawable.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        private Bindable<Skin> skin;
+
+        /// <summary>
+        /// Whether fallback to default skin should be allowed if the custom skin is missing this resource.
+        /// </summary>
+        private readonly bool allowDefaultFallback;
+
+        /// <summary>
+        /// Create a new
```
Comment (smoogipoo) [inline osu.Game/Skinning/SkinReloadableDrawable.cs]: A new what peppy?! The suspense is killing me!

File: osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs [hunk 6]
```diff
@@ -32,8 +31,12 @@ public abstract class DrawableHitObject : Container, IHasAccentColour
         // Todo: Rulesets should be overriding the resources instead, but we need to figure out where/when to apply overrides first
         protected virtual string SampleNamespace => null;
 
-        protected List<SampleChannel> Samples = new List<SampleChannel>();
-        protected virtual IEnumerable<SampleInfo> GetSamples() => HitObject.Samples;
+        protected SkinnableSound Samples;
+
+        protected virtual IEnumerable<SampleInfo> GetSamples()
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs]: `GetSamples() => HitObject.Samples`, as it originally was.

File: osu.Game.Rulesets.Taiko/Audio/DrumSampleMapping.cs [hunk 7]
```diff
@@ -14,7 +13,9 @@ public class DrumSampleMapping
         private readonly ControlPointInfo controlPoints;
         private readonly Dictionary<double, DrumSample> mappings = new Dictionary<double, DrumSample>();
 
-        public DrumSampleMapping(ControlPointInfo controlPoints, AudioManager audio)
+        public readonly List<SkinnableSound> Drawables = new List<SkinnableSound>();
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Taiko/Audio/DrumSampleMapping.cs]: How about rename this to `Sounds`?

File: osu.Game.Rulesets.Taiko/Audio/DrumSampleMapping.cs [hunk 8]
```diff
@@ -27,20 +28,34 @@ public DrumSampleMapping(ControlPointInfo controlPoints, AudioManager audio)
...
                 {
-                    Centre = s.GetSampleInfo().GetChannel(audio.Sample, "Taiko"),
-                    Rim = s.GetSampleInfo(SampleInfo.HIT_CLAP).GetChannel(audio.Sample, "Taiko")
+                    Centre = addDrawableSound(centre),
+                    Rim = addDrawableSound(rim)
                 };
             }
         }
 
+        private SkinnableSound addDrawableSound(SampleInfo rim)
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Taiko/Audio/DrumSampleMapping.cs]: along with previous comment, rename to `addSound` + rename the `rim` parameter to `sampleInfo`

**Review by smoogipoo — CHANGES_REQUESTED** [review]
This looks pretty alright to me. We're going to want tests for this, and probably commenting for skin's functions, but I'm not too pained about that.

**Comment by smoogipoo** [comment]
screen shot 2018-02-24 at 11 09 30 pm

Known issue? The approach circle is smaller than the circle at the point of being hit.

Textures: https://imgur.com/a/eOdO2

**Comment by peppy** [comment]
The texture is probably too large.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**